Docs/sphinx_documentation/source/Basics.rst

```diff
@@ -2110,39 +2152,56 @@ These loops should usually use :cpp:`i <= hi.x`, not :cpp:`i < hi.x`, when
 defining the loop bounds. If not, the highest index cells will be left out
 of the calculation.
 
-.. _sec:basics:loop:
+.. _sec:basics:parallelfor:
 
-Loop and LoopConcurrent
-=======================
+ParallelFor
+===========
 
 In the examples so far, we have explicitly written out the for loops
-when we iterate over a :cpp:`Box` or a range specified by
-:cpp:`Dim3`\ s.  AMReX also provides function templates for writing
-these in a concise way like below,
+when we iterate over a :cpp:`Box.  AMReX also provides function
+templates for writing these in a concise and performance portable way
+like below,
 
 .. highlight:: c++
 
 ::
 
-    Box bx(...);
+    for (MFIter mfi(mfa); mfi.isValid(); ++mfi)
+    {
+      const Box& bx = mfi.validbox();
+      Array4<Real> const& a = mfa[mfi].array();
+      Array4<Real const> const& b = mfb[mfi].const_array();
+      Array4<Real const> const& c = mfc[mfi].const_array();
+      ParallelFor(bx, [=] AMREX_GPU_DEVICE (int i, int j, int k)
+      {
+        a(i,j,k) += b(i,j,k) * c(i,j,k);
+      });
+    }
+
+Here, :cpp:`PareallelFor` takes two arguments.  The first argument is
+a :cpp:`Box` specifying the iteration index space, and the second
+argument is a C++ lambda function that works on cell :cpp:`(i,j,k)`.
+Variables a, b and c in the lambda function are captured by value from
+the enclosing scope.  There are other versions of :cpp:`ParalleFor`,
 
-    // Loop over Box
-    Loop(bx, [] (int i, int j, int k) { ... });
+.. highlight:: c++
 
-    // Loop over Dim3 ranges
-    const Dim3 lo = lbound(bx);
-    const Dim3 hi = ubound(bx); // Inclusive upper bound
-    Loop(lo, hi, [] (int i, int j, int k) { ... });
+::
 
-    // 4D loop
-    Loop(bx, numcomps, [] (int i, int j, int k, int comp) { ... });
+    // 1D for loop
+    ParallelFor(N, [=] AMREX_GPU_DEVICE (int i) { ... });
 
-    // Give hint to compilers that loops can run concurrently
-    LoopConcurrent(bx, [] (int i, int j, int k) { ... });
+    // 4D for loop
+    ParallelFor(box, numcomps,
+                [=] AMREX_GPU_DEVICE (int i, int j, int k, int n) { ... });
 
-Note that, unlike :cpp:`ParalleFor` that will be introduced in section
-:ref:`sec:gpu:for`, they do not launch GPU kernels to do the work,
-although they can be used in GPU kernels.
+The code above is performance portable.  It works with and without
+GPU.  When AMReX is built with GPU support, AMREX_GPU_DEVICE indicates
+that the lambda function is a device function and :cpp:`ParallelFor`
+launches a GPU kernel to do the work.  When it is built without GPU
+support, AMREX_GPU_DEVICE has no effects whatsoever.  More details on
+:cpp:`ParalleFor` that will be presented in section
+:ref:`sec:gpu:for`.
```

Comment:
Based on this discussion, I think some people will get the idea that `ParallelFor` does OpenMP. Maybe we should add a note that this is not the case, and that OpenMP is meant to handled in the outer `MFIter` loops.

Reply:
sure

Reply:
What do you think about the current version?  Please feel free to edit directly.